File: src/interfaces/api/address.ts
/*
 *
 *     Copyright (C) 2018 ConnectUS
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */


import IPointAPI from "./point";
import IAddress from "../internal/address";

export default class IAddressAPI {
    street?: string;
    city?: string;
    province?: string;
    country?: string;
    postal_code?: string;
    apt_number?: string;
    geojson?: IPointAPI;

    public constructor(address: IAddress) {
        if (address == undefined) return;
        this.street = address.street;
        this.city = address.city;
        this.province = address.province;
        this.country = address.country;
        this.postal_code = address.postal_code;
        this.apt_number = address.apt_number;
        this.geojson = IPointAPI.constructPointFromInternal(address.geojson);
    }
}

File: src/interfaces-api/address.go
package interfaces_api

import (
	"encoding/json"
)

type IAddressAPI struct {
	Street        string    `json:"street"`
	City          string    `json:"city"`
	Province      string    `json:"province"`
	Country       string    `json:"country"`
	PostalCode    string    `json:"postal_code"`
	AptNumber     string    `json:"apt_number"`
	GeoJSON       IPointAPI `json:"geojson"`
}

func (address *IAddressAPI) UnmarshalJSON(data []byte) error {
	println(string(data))
	var v []string
	if err := json.Unmarshal(data, &v); err != nil {
		return err
	} // TODO in general
	address.Street = v[1]
	address.City = v[2]
	address.Province = v[3]
	address.Country = v[4]
	address.PostalCode = v[5]
	address.AptNumber = v[6]

	return nil
}